scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala

```diff
@@ -96,20 +96,24 @@ class KRouter[A](
       network.ping(recordToPing, Ping(uuidSource(), config.nodeRecord)).runToFuture.onComplete {
         case Success(_) =>
           // if it does respond, it is moved to the tail and the other node record discarded.
-          bucket.touch(recordToPing.id)
-          debug(s"Ping to ${recordToPing.id.toHex} in bucket $iBucket successful.")
-          info(
-            s"Moving ${recordToPing.id} to head of bucket $iBucket. Discarding (${nodeRecord.id.toHex}, $nodeRecord) as routing table candidate."
-          )
-          debug(s"iBucket($iBucket) = $bucket")
+          if (bucket.size >= config.k) {
+            bucket.touch(recordToPing.id)
```

Comment:
note two things:
1. If I am correct, we don't want that `bucket.size > config.k`, if two methods call `add` concurrently I believe, as @krcz pointed above, that you could add more than k elements as `touch` adds the element when it is not present.
2. note that `touch` is a pure method. So, this call is creating a new updated TimeSet that is not stored anywhere. Hence, the effect is "lost"